src/stan/variational/advi.hpp

```diff
@@ -374,13 +553,29 @@ namespace stan {
       BaseRNG& rng_;
       int n_monte_carlo_grad_;
       int n_monte_carlo_elbo_;
-      double eta_adagrad_;
       int eval_elbo_;
       int n_posterior_samples_;
       std::ostream* print_stream_;
       std::ostream* out_stream_;
       std::ostream* diag_stream_;
 
+      double circ_buff_median_(const boost::circular_buffer<double>& cb) const {
+        // FIXME: naive implementation; creates a copy as a vector
+        std::vector<double> v;
+        for (boost::circular_buffer<double>::const_iterator i = cb.begin();
+              i != cb.end(); ++i) {
+          v.push_back(*i);
+        }
+
+        size_t n = v.size() / 2;
+        std::nth_element(v.begin(), v.begin()+n, v.end());
+        return v[n];
+      }
+
+      double rel_difference_(double prev, double curr) const {
```

Comment:
It's better to allocate ahead of time and assign if you know the size
rather than repeatedly calling push_back.

I didn't know about that n_th_element function --- nifty!

- Bob

On Aug 26, 2015, at 1:10 PM, Alp Kucukelbir <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> >        int eval_elbo_;
> >        int n_posterior_samples_;
> >        std::ostream* print_stream_;
> >        std::ostream* out_stream_;
> >        std::ostream* diag_stream_;
> >  
> > +      double circ_buff_median_(const boost::circular_buffer<double>& cb) const {
> > +        // FIXME: naive implementation; creates a copy as a vector
> > +        std::vector<double> v;
> > +        for (boost::circular_buffer<double>::const_iterator i = cb.begin();
> > +              i != cb.end(); ++i) {
> > +          v.push_back(*i);
> > +        }
> > +
> > +        size_t n = v.size() / 2;
> > +        std::nth_element(v.begin(), v.begin()+n, v.end());
> > +        return v[n];
> > +      }
> > +
> > +      double rel_difference_(double prev, double curr) const {
> 
> let's move these functions to stan-math in a separate pull request after merging
> 
> —
> Reply to this email directly or view it on GitHub.
>